File: distributed-rate-limiter/src/rate-limiter/token_bucket.ts
import { RateLimiter, tokenBucketConfig, tokenBucketResponse } from './typeModel';
import { redisClient } from '../redis/index';
import * as _ from '../constants';

/**
 * 令牌桶限流器實現
 * rate: number; // 補充令牌速率每秒
 * capacity: number; // 桶容量
 * tokens: number; // 當前令牌數量
 * lastRefill: number; // 上次補充令牌的時間戳 秒
 */
export default class TockenBucket implements RateLimiter {
    async run(key: string): Promise<tokenBucketResponse> {
        const result = await (redisClient as any)[_.TOKEN_BUCKET](1, key, '');
        return Boolean(result);
    }

    async set(key: string, config: tokenBucketConfig): Promise<boolean> {
        const result = redisClient.hset(key, 'rate', config.rate, 'capacity', config.capacity, 'tokens', config.capacity, 'lastRefill', Math.floor(Date.now() / 1000));
        return Boolean(result);
    }

    async delete(key: string): Promise<boolean> {
        const result = await redisClient.del(key);
        return Boolean(result);
    }
}

File: distributed-rate-limiter/src/redis/lua/token_bucket.lua
local key = KEYS[1]

local arr = redis.call('HGETALL', key)
-- 沒有刪除，直接通過
if #arr == 0 then
    return 1
end

-- 將數組轉換為對象
local obj = {}
for i = 1, #arr, 2 do
    obj[arr[i]] = tonumber(arr[i + 1])
end

if obj['tokens'] > 0 then
    -- 如果令牌數量大於0，則允許通過
    redis.call('HSET', key, 'tokens', obj['tokens'] - 1)
    return 1
else
    -- 如果令牌數量為0，則補充令牌
    -- [秒, 微秒]
    local now = tonumber(redis.call('TIME')[1])  
    local refillToken = obj['rate'] * (now - obj['lastRefill'])
    local newTokens = math.min(obj['capacity'], obj['tokens'] + refillToken)
    if newTokens > 0 then
        -- 更新令牌數量和最後補充時間
        redis.call('HSET', key, 'tokens', newTokens - 1, 'lastRefill', now)
        return 1
    end
    return 0
end

